File: CSharp/HtmlWalker/FormattedCloneWalker.cs
using HtmlWalker.Platform;
using HtmlWalker.Tag.Html;
using HtmlWalker.Tag.Util;

namespace HtmlWalker
{
    /// <summary>
    ///   This classes clones and does simple formatting.
    /// </summary>
    public class FormattedCloneWalker : CloneWalker
    {
        // TODO: Make <!DOCTYPE html> a new type instead of comment

        private ITag CurrentOwner()
        {
            var owner = currentTag;
            while (owner != null && owner.Name == "#skip")
                owner = owner.Owner;
            return owner;
        }

        public FormattedCloneWalker(WalkerPlatform.WalkerDocument document) : base(document) { }

        protected override ITag HandleTag<T>(T adapter)
        {
            var tag = base.HandleTag(adapter);
            var owner = CurrentOwner();
            if (owner != null && tag.Name != "#skip")
            {
                if (owner.Name == "body")
                    owner.OwnedTags.Add(new Eol(Output.Factory, 2));
                else if (owner.Name == "head")
                    owner.OwnedTags.Add(new Eol(Output.Factory));
            }
            return tag;
        }

        protected override void EndTag()
        {
            bool skipped = (currentTag.Name == "#skip");
            base.EndTag();
            if (!skipped)
            {
                var owner = CurrentOwner();
                if (owner != null)
                {
                    if (owner.Name == "body")
                        owner.OwnedTags.Add(new Eol(Output.Factory, 2));
                    else if (owner.Name == "head")
                        owner.OwnedTags.Add(new Eol(Output.Factory));
                }
            }
        }

        public override void Start(TagHtml adapter)
        {
            base.Start(adapter);
            if (currentTag.Name != "#skip")
                currentTag.OwnedTags.Add(new Eol(Output.Factory));
        }

        public override void Start(TagHead adapter)
        {
            base.Start(adapter);
            if (currentTag.Name != "#skip")
                currentTag.OwnedTags.Add(new Eol(Output.Factory));
        }

        public override void End(TagHead adapter)
        {
            bool skipped = (currentTag.Name == "#skip");
            base.End(adapter);
            if (!skipped)
                currentTag.OwnedTags.Add(new Eol(Output.Factory));
        }

        public override void Start(TagBody adapter)
        {
            base.Start(adapter);
            if (currentTag.Name != "#skip")
                currentTag.OwnedTags.Add(new Eol(Output.Factory, 2));
        }

        public override void End(TagBody adapter)
        {
            bool skipped = (currentTag.Name == "#skip");
            base.End(adapter);
            if (!skipped)
                currentTag.OwnedTags.Add(new Eol(Output.Factory));
        }

        public override void Handle(Eol adapter)
        {
            var owner = CurrentOwner();
            if (
                owner != null &&
                owner.Name != "#skip" &&
                owner.Name != "body" &&
                owner.Name != "head" &&
                owner.Name != "html"
            )
                base.Handle(adapter);
        }
    }
}

File: Java/src/htmlwalker/FormattedCloneWalker.java
/**
 * 
 */
package htmlwalker;

import htmlwalker.exception.HtmlWalkerException;
import htmlwalker.platform.WalkerDocument;
import htmlwalker.tag.html.*;
import htmlwalker.tag.util.*;

/**
 * This classes clones and does simple formatting.
 * 
 * @author Robert R Flores
 */
public class FormattedCloneWalker extends CloneWalker
{

    private ITag currentOwner()
    {
        ITag owner = currentTag;
        while (owner != null && owner.name() == "#skip")
            owner = owner.owner();
        return owner;
    }

	public FormattedCloneWalker(WalkerDocument document) { super(document); }

	@Override
    protected <T extends ITag> ITag handleTag(T adapter) throws HtmlWalkerException
    {
        ITag tag = super.handleTag(adapter);
        ITag owner = currentOwner();
        if (owner != null && tag.name() != "#skip")
        {
            if (owner.name() == "body")
                owner.ownedTags().add(new Eol(output.factory(), 2));
            else if (owner.name() == "head")
                owner.ownedTags().add(new Eol(output.factory()));
        }
        return tag;
    }

	@Override
    protected void endTag()
    {
        boolean skipped = (currentTag.name() == "#skip");
        super.endTag();
        if (!skipped)
        {
            ITag owner = currentOwner();
            if (owner != null)
            {
                if (owner.name() == "body")
                    owner.ownedTags().add(new Eol(output.factory(), 2));
                else if (owner.name() == "head")
                    owner.ownedTags().add(new Eol(output.factory()));
            }
        }
    }

	@Override
    public void start(TagHtml adapter) throws HtmlWalkerException
    {
        super.start(adapter);
        if (currentTag.name() != "#skip")
        	currentTag.ownedTags().add(new Eol(output.factory()));
    }

	@Override
    public void start(TagHead adapter) throws HtmlWalkerException
    {
        super.start(adapter);
        if (currentTag.name() != "#skip")
        	currentTag.ownedTags().add(new Eol(output.factory()));
    }

	@Override
    public void end(TagHead adapter)
    {
        boolean skipped = (currentTag.name() == "#skip");
        super.end(adapter);
        if (!skipped)
            currentTag.ownedTags().add(new Eol(output.factory()));
    }

	@Override
    public void start(TagBody adapter) throws HtmlWalkerException
    {
        super.start(adapter);
        if (currentTag.name() != "#skip")
        	currentTag.ownedTags().add(new Eol(output.factory(), 2));
    }

	@Override
    public void end(TagBody adapter)
    {
        boolean skipped = (currentTag.name() == "#skip");
        super.end(adapter);
        if (!skipped)
            currentTag.ownedTags().add(new Eol(output.factory()));
    }

	@Override
    public void handle(Eol adapter) throws HtmlWalkerException
    {
        ITag owner = currentOwner();
        if (
            owner != null &&
            owner.name() != "#skip" &&
            owner.name() != "body" &&
            owner.name() != "head" &&
            owner.name() != "html"
        )
            super.handle(adapter);
    }
}
